File: wordladder/wordladder.go
package wordladder

import "log"

type Word struct {
	word string
}

func NewWord(word string) Word {
	return Word{word}
}

func NewWords(words []string) []Word {
	results := []Word{}
	for _, w := range words {
		results = append(results, NewWord(w))
	}
	return results
}

func (w Word) LastChar() Char {
	return NewChar(string(w.word[len(w.word)-1]))
}

func (w Word) FirstChar() Char {
	return NewChar(string(w.word[0]))
}

type Problem struct {
	words []Word
}

func (p Problem) Words() []Word {
	return p.words
}

func NewProblem(words []Word) Problem {
	return Problem{words}
}

type Char struct {
	char string
}

func NewChar(char string) Char {
	if len(char) != 1 {
		log.Fatalf("illegal character: %s", char)
	}
	return Char{char}
}

type Answer struct {
	last  Char
	first Char
}

func NewAnswer(last, first Char) Answer {
	return Answer{last, first}
}

type Solver interface {
	Solve(p Problem) (bool, Answer)
}

func New() Solver {
	return &solverImpl{}
}

type solverImpl struct{}

func (s *solverImpl) Solve(p Problem) (bool, Answer) {
	words := p.Words()
	for i := 0; i < len(words)-1; i++ {
		word := words[i]
		nextWord := words[i+1]
		last := word.LastChar()
		first := nextWord.FirstChar()
		if last != first {
			return false, NewAnswer(last, first)
		}
	}
	return true, Answer{}
}

File: wordladder/wordladder.py
import dataclasses
import sys


@dataclasses.dataclass
class Problem:
    words: list[str]


@dataclasses.dataclass
class Answer:
    chained: bool
    last: str | None = None
    first: str | None = None


def solve(problem: Problem) -> Answer:
    for i in range(len(problem.words) - 1):
        word = problem.words[i]
        next_word = problem.words[i + 1]
        last = last_char(word)
        first = first_char(next_word)
        if last != first:
            return Answer(chained=False, last=last, first=first)
    return Answer(chained=True)


def last_char(word: str) -> str:
    return word[len(word) - 1]


def first_char(word: str) -> str:
    return word[0]


if __name__ == "__main__":
    lines = sys.stdin.readlines()
    words = list(map(lambda s: s.strip(), lines[1:]))
    answer = solve(Problem(words))
    if answer.chained:
        print("Yes")
    else:
        print(f"{answer.last} {answer.first}")
